### ISSUE TITLE
WebXR Device API

### ISSUE THREAD

**Issue by kearwood** [mozilla/standards-positions#218]
## Request for Mozilla Position on an Emerging Web Specification

* Specification Title: WebXR Device API
* Specification or proposal URL: https://www.w3.org/TR/webxr/
* Caniuse.com URL (optional): https://caniuse.com/#feat=webxr
* Bugzilla URL (optional): https://bugzilla.mozilla.org/show_bug.cgi?id=1419190
* Mozillians who can provide input (optional): @kearwood, @larsbergstrom, @philip-lamb, @daoshengmu, @thomasmo

### Other information
The WebXR Device API replaces the now deprecated WebVR API, which was shipped in Firefox 55.  Other UA's will be dropping support of WebVR in the transition to WebXR.  WebXR enables immersive VR and AR experiences within the browser.  Support for WebXR has already landed in frameworks used by many web sites, such as aframe, threejs, and Amazon Sumerian.  Several sites created by Mozilla support WebXR, such as Mozilla Hubs.  We must enable support for WebXR in Firefox in order to continue providing access to VR experiences on the web and to pave the way for future AR capabilities.

WebXR improves on WebVR in ways that reflect Mozilla values, such as reduced fingerprinting for tracking protection.

**Comment by dbaron** [comment]
I'm curious how the current API deals with permissions for the various pieces of sensitive data that might be involved, such as:
* sensor information (e.g., orientation, motion, maybe geolocation)
* control over outputs (e.g., vibration, haptics, notifications)
* data from the user's camera

What classes of potentially-sensitive information can a website get access to through this API, and how does the user grant consent for those things?

I'm also curious how trusted browser UI works in VR.  How does the user understand the origin of the page, etc.

**Comment by kearwood** [comment]
> I'm also curious how trusted browser UI works in VR. How does the user understand the origin of the page, etc.

The results of discussion within the working group on trusted UI is here:

https://immersive-web.github.io/webxr/#trusted-ui

I would expect regular desktop Firefox (Not Firefox Reality) to only support the "non-immersive trusted UI".  In this case, the immersive XR sessions will be paused or stopped before any trusted UI is displayed.  For regular desktop Firefox, this would be essentially requiring the user to take off their headset and look at their separate, physical 2d monitor to interact with the browser's prompt.

For Firefox Reality, it has different UI that allows the user to interact with the browser itself in the headset.  It would also initially require stopping the XR session to interact with trusted UI, with the opportunity to later experiment with UX concepts such as those described in the "Examples of trusted UI" section of the spec.

**Comment by kearwood** [comment]
> I'm curious how the current API deals with permissions for the various pieces of sensitive data that might be involved, such as:
> 
>     * sensor information (e.g., orientation, motion, maybe geolocation)
No sensor information is accessible without first requesting an XR session, which in-turn requires a permission prompt.

Geolocation is not exposed by the WebXR API.  Position and orientation are relative to a "Native Origin", which is not correlated to real-world geolocation:

https://immersive-web.github.io/webxr/#xrspace-native-origin

Position and orientation are reported as a "rigid transform", containing a 4-component quaternion and 3-component translation vector within an XRSpace defined coordinate system, which is in-turn relative to the "Native Origin".

Unlike WebVR, the WebXR api no longer exposes acceleration values; however, those can be inferred by sampling the position and orientation over multiple frames.

> 
>     * control over outputs (e.g., vibration, haptics, notifications)
Vibration, haptics, and notifications are not exposed in the current WebXR API.  The earlier WebVR API allowed haptic pulses to be emitted from controllers, through the "Gamepad Extensions API":

https://w3c.github.io/gamepad/extensions.html

The Firefox WebVR implementation only enumerated controllers capable of haptic feedback for the duration of a WebVR immersive session..  The "WebXR Gamepads Module" requires that such controllers are only enumerated by a function on the XRInputSource object, which can only be accessed after the user accepts the permission prompt to start an immersive XR session.  Such controllers will still be represented with the same Gamepad webidl:

https://immersive-web.github.io/webxr-gamepads-module/
> 
>     * data from the user's camera
Data from the user's camera, such as frame captures and exposure information, is not accessible through WebXR.  It is also possible for a device without any cameras at all to support the WebXR api.  Devices with such cameras are performing the processing at the operating system level, or often with dedicated processors and hardware.  The WebXR api only exposes the result of such calculations as a 4-component quaternion and 3-component vector representing the orientation and position of the headset and controllers in space.

Some headsets allow these cameras to additionally be used as web cams; however, such access is out of scope for the WebXR API.  This could be implemented such that they could be exposed through the WebRTC api; however, we have not done so.
> 
> 
> What classes of potentially-sensitive information can a website get access to through this API, and how does the user grant consent for those things?


Information about the IPD (distance between the eyes) is required for stereoscopic rendering.  The VR hardware has a sensor to read back the user's adjusted IPD value.  This is exposed in quantized form only after the user accepts a prompt to start an XR session.

Information about the height of the user's eyes relative to the floor is accessible only when the XR session is requested from the WebXR API with an explicit parameter; otherwise, the positional tracking will always be relative to the starting position when the session was started.  The height above the floor is also quantized.

Information about the dimensions of the physical space is helpful to prevent users from walking into physical barriers when their view is covered by a VR headset.  The lower level platforms provide visual indications that the user is nearing the boundaries, composited over top of the browser's rendered output.  It is useful; however, for XR experiences to know about these boundaries so that they don't ask users to move outside the boundaries and into harm's way.  The dimensions of these boundaries are only available in quantized form, for XR sessions that are requested with an explicit parameter.

**Comment by adamroach** [comment]
Re-ping of @kearwood, @larsbergstrom, @philip-lamb, @daoshengmu, @thomasmo for input. Also adding @martinthomson and @ekr for input on the privacy and security issues that @dbaron raises above.

**Comment by daoshengmu** [comment]
One of the goals of WebXR API is making up functionalities that are not robust enough in WebVR. The potential fingerprinting risk of WebVR API is always a concern. Therefore, we keep improving this part in WebXR API,  the most important difference from WebXR is we ask UA needs to provide permission UI (https://immersive-web.github.io/webxr/#trusted-ui) to get approval from users before accessing their device information.

This scenario will make users understand these websites want to access their VR/AR device information, and we will give a notification to make users understand what these websites are doing.

WebXR API is like the next version of WebVR API. We need it come out and replace its previous unperfected one.

**Comment by Manishearth** [comment]
As requested by @dbaron I'm going to try and systematically write down the different information exposed by the API.

Note that the use of this API in the first place is gated by a permissions prompt and feature policy.

----

Whether or not you have an XR-capable device attached to the browser is exposed via `isSessionSupported()`, without requiring a permissions prompt. Browsers can choose to add an extra prompt here of the form "do you wish for the page to know about your devices?" however, since it returns a promise. This is largely used for showing "enter VR" buttons.

In a normal immersive XR session, the following information is exposed:

 - The relative locations of your headset, floor, and input devices, and how they evolve over time. Velocity and acceleration are not exposed, nor is an "absolute" notion of location like geolocation
 - The distance between your eyes is computable. This is sometimes configurable based on the device.
 - The floor position can be used to guess user height. Using floor spaces requires an additional feature (which can manifest itself in a change in the permissions prompt).
 - The "bounds" of the user's space can be read with an additional feature. This is typically set by the user on the device for each room they use, and needs to be redrawn every now and then (If I move around too many objects in my living room, my device forgets the original bounds). The shape of this can be used to make guesses about the size/shape of the user's room.
 - The relationship between target ray and grip spaces can be used to identify the device
 - The resolution of the device (can be used to identify the device)

There is no control over haptics/notifications/etc, and no camera input without using `getUserMedia()`.

The module _does_ take over the full device rendering loop, so it's similar to the fullscreen API in that it's possible to spoof a browser in this mode. However, [the concept of trusted UI](https://immersive-web.github.io/webxr/#trusted-ui) exists to require certain things be unspoofable, either by using UA-reserved buttons on a controller, or by having "floating UI" that can be moved around but never dismissed or composited over.

Additional modules (treated as separate specifications) expose further info. Most have their own feature and user consent requirements.

[The Gamepads module](https://immersive-web.github.io/webxr-gamepads-module/) additionally exposes access to the button data from the controllers. Controllers have an "input profile list", specifying which controller it is, with fallbacks. If a UA chooses to expose the maximally accurate input profile, you can typically identify which device the user has. If not, the UA will still be exposing the capabilities in the fallback "generic" profile, which can be used to narrow down which device is being used.

[The AR module](https://immersive-web.github.io/webxr-ar-module/) additionally exposes some extra information about AR devices related to how they do blending (essentially a boolean), which can be used to narrow down which device is being used.

[The hit-test module](https://immersive-web.github.io/hit-test/) enables raycasting against the real world. It requires an additional permission, but it exposes details about the real world geometry.

[The anchors module](https://immersive-web.github.io/anchors/) by itself doesn't really do more here, but when paired with hit-test it can be used to track how real-world objects move over time. It requires an additional permission.

[The hands-input module](https://github.com/immersive-web/webxr-hands-input/blob/master/explainer.md) exposes the locations of your hand joints, which is quite "fingerprintable" :wink: It requires an additional permission.

[The lighting estimation module](https://github.com/immersive-web/lighting-estimation/) exposes information on the light distribution in a room. It requires an additional permission.

[The layers module](https://immersive-web.github.io/layers/) doesn't expose anything new as far as I can tell, but it also requires an additional permission.
